### ISSUE TITLE
Strange preloading bug in Ecto 3.5.0-rc.0

### ISSUE THREAD

**Issue by sasa1977** [elixir-ecto/ecto#3415]
### Environment

* Elixir version (elixir -v): 1.10
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): PostgreSQL 12.4
* Ecto version (mix deps): 3.5.0-rc.0
* Database adapter and version (mix deps): postgrex 0.15.5
* Operating system: Ubuntu

### Current behavior

While testing Ecto 3.5.0-rc.0 I've noticed a strange preloading bug. In some cases, a list of multiple preloads sets the association fields to `nil` even though the corresponding rows exist in the database. Here's a demonstration:

```elixir
process = Repo.get!(Process, process_id)

IO.inspect(process.patient)
# #Ecto.Association.NotLoaded<association :patient is not loaded>

IO.inspect(process.encounter)
# #Ecto.Association.NotLoaded<association :encounter is not loaded>

IO.inspect(Repo.preload(process, [:encounter]).encounter)
# %Encounter{...}

IO.inspect(Repo.preload(process, [:patient, :encounter]) |> Map.take([:patient, :encounter]))
# %{encounter: nil, patient: nil}
```

In other words, each association is correctly loaded if it's preloaded individually, but neither is fetched when both are being preloaded.

Looking at Ecto logs, it seems that correct database queries are issued.
Both, `:patient` and `:encounter` associations are defined in the `Process` schema as `belongs_to`. I've also tried the same experiment on another schema with `has_one` associations, and that worked fine.

**Comment by josevalim** [comment]
I cannot reproduce this. I have added this test to our suite:

```
test "preload multiple belongs_to" do
    %User{id: uid} = TestRepo.insert!(%User{name: "foo"})
    %Post{id: pid} = TestRepo.insert!(%Post{title: "1"})
    %Comment{id: cid} = TestRepo.insert!(%Comment{post_id: pid, author_id: uid})

    comment = TestRepo.get!(Comment, cid)
    comment = TestRepo.preload(comment, [:author])
    assert comment.author.id == uid

    comment = TestRepo.preload(comment, [:author, :post])
    assert comment.author.id == uid
    assert comment.post.id == pid
  end
```

And it works fine! You can try playing with the test by cloning both `ecto` and `ecto_sql` in the same directory and then from `ecto_sql` you run:

     ECTO_ADAPTER=pg mix test ../ecto/integration_test/cases/preload.exs

**Comment by sasa1977** [comment]
So, I forgot to mention one important piece of information: this happens on a dynamic repo.

After going through the depts of ecto & ecto_sql, I was able to narrow it down to [maybe_pmap](https://github.com/elixir-ecto/ecto/blob/d173e0f74c54dc03dfd815f4ad98e491cf340516/lib/ecto/repo/preloader.ex#L123-L132). I first established that if I disable parallel preloads, everything works.

Then, I reverted back to parallel preloading, and [inside the task function](https://github.com/elixir-ecto/ecto/blob/d173e0f74c54dc03dfd815f4ad98e491cf340516/lib/ecto/repo/preloader.ex#L127) I hardcoded:

```elixir
MyRepo.put_dynamic_repo(repo_name)
Ecto.Adapters.SQL.Sandbox.allow(MyRepo, parent_pid, self())
```

With this, parallel preloading works. Note that I had to do both, set the dynamic repo, and setup sandbox allowance. Not sure what the proper fix should be, but hopefully this gives you some useful information. Let me know if I can assist further.

**Comment by josevalim** [comment]
If you don’t set allowances, it raises an error? Because it should be using the mechanism that sets allowances based on the task caller

**Comment by sasa1977** [comment]
Good question! If allowance is not set it returns nothing but doesn't raise. Which reminded me to double check something else, and I've noticed that sandbox mode was not set to manual. When I do that then everything is working as expected, so it seems the problem was in my code after all, right? The funny thing is that the same code works fine with Ecto 3.4. Has perhaps the default sandbox mode changed?

**Comment by josevalim** [comment]
@sasa1977 we did change the preloader. The only changes to the sandbox were the ones related to dynamic repo. :) But did the dynamic parallel preload work on v3.4? Because I wouldn't have expected it to.

**Comment by sasa1977** [comment]
> But did the dynamic parallel preload work on v3.4?

Not sure if dynamic parallel preload worked, but all tests are passing on 3.4.6 :-) We're using a copy of the sandbox from 3.5, so that hasn't changed when I updated ecto_sql. Was perhaps parallel preload not used by default before 3.5?

**Comment by josevalim** [comment]
The default is the same. Can you please try checking out [this commit](https://github.com/elixir-ecto/ecto/commit/e75ae5565894973541fdcad43583c6428144c27a#diff-9dab1b1adef58f372bcbdd976c02f663) and let us know if it fails or pass?

**Comment by sasa1977** [comment]
I first tried with 3.4.6 and can confirm that preloads are done in parallel, and everything works correctly. Same thing for the commit you linked. I went on to do manual bisection, and I established that [this change](https://github.com/elixir-ecto/ecto/commit/95c1101f94e26a39ae4c5bebb28ed660fa4bfbf9#diff-9dab1b1adef58f372bcbdd976c02f663L130) breaks the behaviour for me. If I manually return the line 130 it works again.

**Comment by josevalim** [comment]
I double checked DBConnection and that line is necessary if you are not running in `:manual` mode. So the question is indeed why you are not running on manual or {shared, _}. If you run in automatic, it means that each process run on its own transaction, including those tasks. In any case, I will add the line back for now. Thank you so much for bisecting!

**Comment by sasa1977** [comment]
> So the question is indeed why you are not running on manual or {shared, _}

That was an error on our side :-)
